### ISSUE TITLE
compile -v doesn't work for sbt --client

### ISSUE THREAD

**Issue by eed3si9n** [sbt/sbt#5869]
I noticed this while testing https://github.com/sbt/sbt-launcher-package/pull/328

## steps

```
$ time SBT_NATIVE_CLIENT=true  ~/work/sbt-modules/sbt-launcher-package/target/universal/stage/bin/sbt compile -v
```

## problem

```
$ time SBT_NATIVE_CLIENT=true  ~/work/sbt-modules/sbt-launcher-package/target/universal/stage/bin/sbt compile -v
[process_args] java_version = '8'
[sbt_options] declare -a sbt_options='([0]="-Xms2048M" [1]="-Xmx2048M" [2]="-Xss2M")'
[debug] running native client
# Executing command line:
/Users/eed3si9n/work/sbt-modules/sbt-launcher-package/target/universal/stage/bin/sbtn-x86_64-apple-darwin
compile
-v

[info] entering *experimental* thin client - BEEP WHIRR
[info] server was not detected. starting an instance
[info] terminate the server with `shutdown`
> compile -v
[error] Expected whitespace character
[error] Expected '/'
[error] compile -v
[error]         ^
[info] compile -v completed
[error] Total time: 0 s, completed Sep 19, 2020 10:12:44 PM
SBT_NATIVE_CLIENT=true  compile -v  0.24s user 0.27s system 5% cpu 8.551 total
```

## expectation

`-v` is should be used as the argument to `sbt --server`, but should be ignored as a command.

## notes

If I flip the order to `-v compile` then it works.

**Comment by eatkins** [comment]
I'm not sure I agree that this should be a supported use case. What if an input task wants to have a `-v` argument?

**Comment by eatkins** [comment]
I guess the workaround for such an input task would be `sbtn -- foo -v` if the thin client argument parser supported `--`. I'm coming at this from the java perspective where all the `-` arguments come before the class name and everything after the class name is treated as a command argument regardless of whether they contain a `-`.

**Comment by eed3si9n** [comment]
> I'm not sure I agree that this should be a supported use case. What if an input task wants to have a `-v` argument?

Shouldn't `sbtn` be a drop-in (or "bug-compatible") replacement for `sbt`?
`-v` is filtered out at bash level, so afaik it can appear at any position.
If an input task needs to take an argument, we currently do so using double quotes? However I'm not sure how good the quotation handling is tbh.

**Comment by eatkins** [comment]
I'm really arguing that the sbt bash script is broken. By stripping the trailing -v, it makes command invocation inconsistent when invoked from the sbt shell vs the bash script.

**Comment by eatkins** [comment]
To elaborate, `sbt compile -v` makes it looks like `compile` accepts a `-v` flag, but it doesn't. I find that confusing.

**Comment by eed3si9n** [comment]
Some tools (like bazel) are very specific about where I can put which options for what subcommand, and it's difficult for newcomers. I kind of like how sbt lets me put options wherever. I do like `--` used as emulation of stdin or to mark "from here it's actually arguments".

**Comment by eed3si9n** [comment]
Here's another example where sbt doesn't care about option-argument ordering. Traditionally sbt used `-` as early command. This means you could write:

```
$ sbt -warn compile
$ sbt compile -warn
```

and it would sort out `-warn` and run it before metabuild loading.

**Comment by eatkins** [comment]
I strongly dislike this behavior and I think it leads to more confusion in the long run. I much prefer my tools to be consistent. There shouldn't be different behavior if I invoke a command from the sbt shell or the terminal shell which is what this does.

**Comment by eatkins** [comment]
If we reworked the sbt parser so that `compile -warn` set the log level to warn and then ran compile, I would be fine with it. It's the inconsistency that bothers me.

**Comment by eed3si9n** [comment]
> I strongly dislike this behavior and I think it leads to more confusion in the long run. I much prefer my tools to be consistent. There shouldn't be different behavior if I invoke a command from the sbt shell or the terminal shell which is what this does.

By the time the user is in sbt shell, Bash script or metabuild loading has finished. So it doesn't make sense to set either `-v` or `-warn` from sbt shell.

So the question is: Should we force users to group batch-only options earlier than arguments? I think there are different schools of thoughts on it? Making change now might result to another set of confused users especially given the state of the parser error message like in this report:

```
[error] Expected whitespace character
[error] Expected '/'
[error] compile -v
[error]         ^
```

**Comment by eatkins** [comment]
The way it could be handled is that we could detect in the bash script if they put a dash argument after the first non-dash argument and print a warning at the bash script level.

**Comment by eatkins** [comment]
>  So it doesn't make sense to set either -v or -warn from sbt shell.

I agree that `-warn compile` makes no sense but `compile -warn` does. One nice way to handle this would be to allow an optional trailing log argument to any task in the parser. So, for example, you could run `test -d` and we would set the log level to debug before evaluating the command and reset it after.

**Comment by eed3si9n** [comment]
Closing this for now.
